ext/standard/array.c

```diff
@@ -3511,11 +3511,21 @@ PHP_FUNCTION(array_slice)
 	if (HT_IS_PACKED(Z_ARRVAL_P(input)) &&
 	    (!preserve_keys ||
 	     (offset == 0 && HT_IS_WITHOUT_HOLES(Z_ARRVAL_P(input))))) {
+
+		zend_bool can_skip = offset > 0 && HT_IS_WITHOUT_HOLES(Z_ARRVAL_P(input));
 		zend_hash_real_init_packed(Z_ARRVAL_P(return_value));
 		ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(return_value)) {
 			ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(input), entry) {
 				pos++;
 				if (pos <= offset) {
+					if (can_skip) {
+						/* Optimization: Instead of advancing offset times over a packed array without holes, */
+						/* advance just once (useful when offset is a million). */
+						/* The above code checked that the offset was between 0 and the length of the array (inclusive). */
+						/* TODO: Better abstraction to advance hash pointer in ZEND_HASH_FOREACH? */
+						_p += (offset - pos);
+						pos = offset;
+					}
```

Comment:
I'd prefer to structure this code differently: Split out the actual modification after the loop (using entry) and then convert into an if-else that either does the loop, or directly addresses the right bucket.

Mixing both is pretty non-obvious and will also perform an unnecessary can_skip check on each iteration.

Reply:
Do you mean getting rid of the ZEND_HASH_FOREACH_VAL macro entirely and using two separate for loops?

> and then convert into an if-else that either does the loop, or directly addresses the right bucket.

I assume you mean something along these lines:

```
					if (can_skip) {
						/* Optimization: Instead of advancing offset times over a packed array without holes, */
						/* advance just once (useful when offset is a million). */
						/* The above code checked that the offset was between 0 and the length of the array (inclusive). */
						/* TODO: Better abstraction to advance hash pointer in ZEND_HASH_FOREACH? */
						_p += (offset - pos);
						pos = offset;
					} else {
						while (_p != end && pos < offset) {
							zval *_z = &_p->val;
							if (UNEXPECTED(Z_TYPE_P(_z) == IS_UNDEF)) {
								continue;
							}
							pos++;
						}
					}
```